interop-scoring/main.js

```diff
@@ -0,0 +1,506 @@
+  '/webcodecs/videoDecoder-codec-specific.https.any.worker.html?vp8',
+  '/webcodecs/videoDecoder-codec-specific.https.any.worker.html?vp9',
+  '/webcodecs/videoDecoder-codec-specific.https.any.worker.html?av1',
+  '/webcodecs/videoFrame-construction.any.html',
+  '/webcodecs/videoFrame-construction.crossOriginSource.sub.html',
+  '/webcodecs/videoFrame-construction.window.html',
+  '/webcodecs/videoFrame-serialization.crossAgentCluster.https.html',
+  '/webcodecs/videoFrame-serialization.crossAgentCluster.https.html',
+  '/webcodecs/temporal-svc-encoding.https.any.html?h264',
+  '/webcodecs/temporal-svc-encoding.https.any.html?vp8',
+  '/webcodecs/temporal-svc-encoding.https.any.html?vp9',
+  '/webcodecs/temporal-svc-encoding.https.any.worker.html?h264',
+  '/webcodecs/temporal-svc-encoding.https.any.worker.html?vp8',
+  '/webcodecs/temporal-svc-encoding.https.any.worker.html?vp9',
+  '/webcodecs/videoFrame-serialization.crossAgentCluster.https.html',
+  '/webcodecs/videoFrame-serialization.crossAgentCluster.https.html',
+  '/webcodecs/videoFrame-serialization.crossAgentCluster.https.html',
+  '/webcodecs/full-cycle-test.https.any.html?av1',
+  '/webcodecs/full-cycle-test.https.any.html?h264_annexb',
+  '/webcodecs/full-cycle-test.https.any.html?h264_avc',
+  '/webcodecs/full-cycle-test.https.any.html?vp9_p0',
+  '/webcodecs/full-cycle-test.https.any.html?vp9_p2',
+  '/webcodecs/full-cycle-test.https.any.worker.html?av1',
+  '/webcodecs/full-cycle-test.https.any.worker.html?h264_annexb',
+  '/webcodecs/full-cycle-test.https.any.worker.html?h264_avc',
+  '/webcodecs/full-cycle-test.https.any.worker.html?vp9_p0',
+  '/webcodecs/full-cycle-test.https.any.worker.html?vp9_p2',
+  '/webcodecs/full-cycle-test.https.any.html?vp8',
+  '/webcodecs/full-cycle-test.https.any.worker.html?vp8',
+  // interop-2023-webcomponents
+  '/shadow-dom/focus/focus-shadowhost-display-none.html',
+  '/custom-elements/form-associated/ElementInternals-labels.html',
+  '/custom-elements/form-associated/ElementInternals-setFormValue.html',
+  '/custom-elements/form-associated/ElementInternals-validation.html',
+  '/custom-elements/form-associated/form-disabled-callback.html',
+]);
+
+
+// Calculate interop score (passing in all browsers) for a category
+// after tracking the category's scores for each browser.
+function aggregateInteropTestScores(interopScores, numBrowsers) {
+  if (interopScores.length === 0) return 0;
+  let aggregateScore = 0;
+  for (const [, results] of interopScores) {
+    let subtestsAllPassing = 0;
+    for (const [, subtestNumPassing] of results) {
+      // A subtest counts toward the interop score if every browser passed the subtest.
+      // In other words, the number of passes should match the number of browsers.
+      if (subtestNumPassing === numBrowsers) {
+        subtestsAllPassing += 1;
+      }
+    }
+    aggregateScore += Math.floor(1000 * subtestsAllPassing / results.size);
```

Comment:
I think `results.size` here is the number of unique subtests, which could be larger than the number of subtests seen in any browser. In that case I think it will be impossible to get an interop score of 100%, right?

I wonder if we should instead do this: if all of the subtests are present in all runs, then the interop score is the number of tests that passed in all runs, just like the code is now. Otherwise, the interop score is the lowest score of any of the runs. That might be too high, but for such a test we really can't reason about the tests that only appear in some browser. WDYT?

A further simplification would be to just use the lowest score of any browser. That's the "otherwise" branch above. If the difference between doing that and the more complicated approach is small, we can suggest that to the interop team, as it would be easier to understand.

Documenting at the top of the file how this works will be great for understanding it in the future without reading all of the code.

Reply:
> I think results.size here is the number of unique subtests, which could be larger than the number of subtests seen in any browser. In that case I think it will be impossible to get an interop score of 100%, right?

This is true - if a browser has a test result listed that another browser does not have, then that test will not be given interoperability score credit.

> I wonder if we should instead do this: if all of the subtests are present in all runs, then the interop score is the number of tests that passed in all runs, just like the code is now. Otherwise, the interop score is the lowest score of any of the runs. That might be too high, but for such a test we really can't reason about the tests that only appear in some browser. WDYT?

This unfortunately swings the problem in the opposite direction. In these cases, the interop score will be skewed upward, even to the point that there could be scenarios where the interop score would display a number that is higher than any percentage browser scores show.

For example:
|Test | Chrome | Firefox | Safari |
|----|---------|---------|------|
| test1 | Pass | Pass | Pass |
| test2 | Fail | Fail | Pass |
| test3 | Pass | Pass | Pass |
| test4 | Pass | Missing | Pass |

In the above test suite, this would currently be scored as:
```
Chrome: 75%, Firefox: 66%, Safari: 100%, Interop: 50%
```

However, if we changed it so the score was averaged without taking into account missing tests and using the lowest test count, the results would look like this:
```
Chrome: 75%, Firefox: 66%, Safari: 100%, Interop: 75%
```

I think it would be a bit more confusing if we had a situation where the interop score was higher than a browser's overall score. Neither of these options are perfect, and it is unfortunate what missing test results can do to this score.